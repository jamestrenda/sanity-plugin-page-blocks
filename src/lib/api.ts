export const apiVersion = '2025-04-13'
