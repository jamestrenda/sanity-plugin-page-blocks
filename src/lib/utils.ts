/**
 * Merges user-defined field groups with default field groups.
 *
 * If a user-defined group has the same `name` as a default group, the properties of the default group
 * will be overwritten by the user-defined group while keeping any properties not explicitly provided by the user.
 * Any additional user-defined groups that do not exist in the default groups will be appended to the result.
 *
 * @template T - The shape of the group object, which must include a `name` property.
 * @param {T[]} defaultGroups - The default field groups provided by the plugin.
 * @param {Partial<T>[]} [userGroups] - Optional user-defined field groups that override or extend the defaults.
 * @returns {T[]} - The merged array of field groups, ensuring defaults are respected while allowing user overrides.
 *
 * @example
 * const defaultGroups = [
 *   { name: 'main', title: 'Content', default: true }
 * ];
 *
 * const userGroups = [
 *   { name: 'main', icon: ComposeIcon, default: false }
 * ];
 *
 * const merged = mergeGroups(defaultGroups, userGroups);
 * console.log(merged);
 * // Output:
 * // [
 * //   { name: 'main', title: 'Content', default: false, icon: ComposeIcon }
 * // ]
 */
export function mergeGroups<T extends {name: string}>(
  defaultGroups: T[],
  userGroups?: Partial<T>[],
): T[] {
  return [
    ...defaultGroups.map((defaultGroup) => {
      const userGroup = userGroups?.find((group) => group.name === defaultGroup.name)
      return userGroup ? ({...defaultGroup, ...userGroup} as T) : defaultGroup
    }),
    ...((userGroups?.filter(
      (group) => !defaultGroups.some((defaultGroup) => defaultGroup.name === group.name),
    ) as T[]) ?? []),
  ]
}

/**
 * Determines whether a given field should be hidden.
 *
 * Fields are **visible by default**, meaning they are only considered hidden
 * if they are explicitly set to `false`. If the field is `undefined` or an
 * object, it is considered visible.
 *
 * @template T - The expected object type of the field when it is visible.
 * @param field - The field value, which can be an object (visible), `false` (hidden), or `undefined` (visible by default).
 * @returns `true` if the field is explicitly set to `false` (hidden), otherwise `false`.
 *
 * @example
 * ```ts
 * const title = { components: 'SomeComponent' };
 * console.log(isFieldHidden(title)); // false (visible)
 *
 * console.log(isFieldHidden(false)); // true (hidden)
 *
 * console.log(isFieldHidden(undefined)); // false (visible)
 * ```
 *
 * @since 1.3.0
 */
export function isFieldHidden<T>(field: T | false | undefined): field is false {
  return field === false
}
