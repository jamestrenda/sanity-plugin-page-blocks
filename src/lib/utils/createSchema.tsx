import React from 'react'
import {FieldDefinition, FieldGroupDefinition, SchemaTypeDefinition} from 'sanity'

import {SchemaBaseFields} from '../../types'

/**
 * Merges user-defined field groups with default field groups.
 *
 * If a user-defined group has the same `name` as a default group, the properties of the default group
 * will be overwritten by the user-defined group while keeping any properties not explicitly provided by the user.
 * Any additional user-defined groups that do not exist in the default groups will be appended to the result.
 *
 * @param {FieldGroupDefinition[]} defaultGroups - The default field groups provided by the plugin.
 * @param {FieldGroupDefinition[]} [userGroups] - Optional user-defined field groups that override or extend the defaults.
 * @returns {FieldGroupDefinition[]} - The merged array of field groups, ensuring defaults are respected while allowing user overrides.
 *
 * @example
 * const defaultGroups = [
 *   { name: 'main', title: 'Content', default: true }
 * ];
 *
 * const userGroups = [
 *   { name: 'main', icon: ComposeIcon, default: false }
 * ];
 *
 * const merged = mergeGroups(defaultGroups, userGroups);
 * console.log(merged);
 * // Output:
 * // [
 * //   { name: 'main', title: 'Content', default: false, icon: ComposeIcon }
 * // ]
 */
export function mergeGroups(
  defaultGroups: FieldGroupDefinition[],
  userGroups?: FieldGroupDefinition[],
): FieldGroupDefinition[] {
  return [
    ...defaultGroups.map((defaultGroup) => {
      const userGroup = userGroups?.find((group) => group.name === defaultGroup.name)
      return userGroup ? {...defaultGroup, ...userGroup} : defaultGroup
    }),
    ...(userGroups?.filter(
      (group) => !defaultGroups.some((defaultGroup) => defaultGroup.name === group.name),
    ) ?? []),
  ]
}

/**
 * Determines whether a given field should be hidden.
 *
 * Fields are **visible by default**, meaning they are only considered hidden
 * if they are explicitly set to `false`. If the field is `undefined` or an
 * object, it is considered visible.
 *
 * @template T - The expected object type of the field when it is visible.
 * @param field - The field value, which can be an object (visible), `false` (hidden), or `undefined` (visible by default).
 * @returns `true` if the field is explicitly set to `false` (hidden), otherwise `false`.
 *
 * @example
 * ```ts
 * const title = { components: 'SomeComponent' };
 * console.log(isFieldHidden(title)); // false (visible)
 *
 * console.log(isFieldHidden(false)); // true (hidden)
 *
 * console.log(isFieldHidden(undefined)); // false (visible)
 * ```
 *
 * @since 1.3.0
 */
export function isFieldHidden<T>(field: T | false | undefined): field is false {
  return field === false
}

/** Type-safe utility to filter out hidden fields */
// export function getVisibleFields<T extends FieldDefinition>(
//   fields: T[],
//   options: Record<string, any>,
// ): T[] {
//   return fields.filter(({name}) => !isFieldHidden(options?.[name as keyof typeof options]))
// }

export function getVisibleFields<T extends FieldDefinition, O extends Record<string, unknown>>(
  fields: T[],
  options: O,
): T[] {
  return fields.filter(({name}) => !isFieldHidden(options?.[name as keyof O]))
}

export function createFieldConfig<
  T extends {
    components?: T['components']
    fieldset?: FieldDefinition['fieldset']
    group?: FieldDefinition['group']
  },
>(options: T | undefined) {
  return {
    components: options?.components ?? undefined,
    fieldset: options?.fieldset ?? undefined,
    group: options?.group ?? undefined,
  }
}

/** Utility to generate a schema base structure */
// export function createSchema<T extends SchemaBaseFields>(
//   config: T & SchemaTypeDefinition
// ): SchemaTypeDefinition
export function createSchema<T extends Record<string, unknown> & SchemaBaseFields>({
  name,
  title,
  icon,
  fields,
  options,
}: {
  name: string
  title: string
  icon: () => React.JSX.Element
  fields: FieldDefinition[]
  options?: T
}): SchemaTypeDefinition {
  const groups = mergeGroups([], options?.groups)
  const visibleFields = options ? getVisibleFields(fields, options) : fields

  return {
    name: options?.name ?? name,
    title,
    type: 'object',
    fieldsets: [...(options?.fieldsets ?? [])],
    groups,
    icon,
    preview: options?.preview ?? undefined, // TODO: make a default preview
    fields: visibleFields,
    components: options?.components,
  }
}
