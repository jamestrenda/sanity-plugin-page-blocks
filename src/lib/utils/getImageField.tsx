import {ObjectDefinition} from 'sanity'

import {CustomImageType} from '../../types'
import {imageField} from '../fields/imageField'
import {imageFields} from '../fields/imageFields'

/**
 * Generates an image field configuration for Sanity schema.
 *
 * This function constructs an image field based on the provided options, allowing customization
 * through additional fields. Since `false` is excluded from `options`, a valid object must be provided.
 *
 * @param {Exclude<CustomImageType, false>} options - Configuration options for the image field.
 *   - Additional `customFields` can be specified to extend the image schema.
 * @returns {ObjectDefinition} - The configured image field.
 */
export function getImageField(options: Exclude<CustomImageType, false>): ObjectDefinition {
  if (!options) {
    return imageField([], options)
  }
  return imageField([...imageFields(options), ...(options?.customFields ?? [])], options)
}
