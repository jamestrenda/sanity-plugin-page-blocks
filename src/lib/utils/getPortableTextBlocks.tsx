import {BlockDefinition, defineArrayMember} from 'sanity'

export const portableTextBlocks = defineArrayMember({
  type: 'block',
  styles: [
    {title: 'Normal', value: 'normal'},
    {title: 'Heading 1', value: 'h1'},
    {title: 'Heading 2', value: 'h2'},
    {title: 'Heading 3', value: 'h3'},
    {title: 'Heading 4', value: 'h4'},
    {title: 'Heading 5', value: 'h5'},
    {title: 'Heading 6', value: 'h6'},
    {title: 'Quote', value: 'blockquote'},
  ],
  lists: [
    {title: 'Bullet', value: 'bullet'},
    {title: 'Number', value: 'number'},
  ],
  // Marks let you mark up inline text in the block editor.
  marks: {
    // Decorators usually describe a single property - e.g. a typographic
    // preference or highlighting by editors.
    decorators: [
      {title: 'Strong', value: 'strong'},
      {title: 'Emphasis', value: 'em'},
      {title: 'Underline', value: 'underline'},
      {title: 'Strike', value: 'strike-through'},
      {title: 'Code', value: 'code'},
    ],
    // Annotations can be any object structure - e.g. a link or a footnote.
    // annotations: [],
  },
})

export function getPortableTextBlocks(options?: Omit<BlockDefinition, 'type' | 'name'>) {
  if (!options) {
    return [portableTextBlocks]
  }
  const {styles, lists, marks} = options

  const result = [
    defineArrayMember({
      ...portableTextBlocks,
      styles: styles ?? portableTextBlocks.styles,
      lists: lists
        ? portableTextBlocks.lists?.filter((list) => lists?.some((l) => l.value === list.value))
        : portableTextBlocks.lists,
      marks: {
        decorators: marks?.decorators
          ? portableTextBlocks.marks?.decorators?.filter((decorator) =>
              marks?.decorators?.some((d) => d.value === decorator.value),
            )
          : portableTextBlocks.marks?.decorators,
        annotations: marks?.annotations ?? portableTextBlocks.marks?.annotations,
      },
    }),
  ]

  return result
}
