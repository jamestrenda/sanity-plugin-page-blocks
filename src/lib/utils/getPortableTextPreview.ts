import {isPortableTextTextBlock, PortableTextBlock} from 'sanity'

/**
 * Generates a preview title from a Portable Text value.
 *
 * @param value - An array of Portable Text blocks.
 * @param blockTitle - A fallback title if no valid text block is found.
 * @returns An object containing `title` and optionally `subtitle`.
 *
 * @public
 */
export function getPortableTextPreview(
  value: PortableTextBlock[] | undefined,
  blockTitle: string,
): {
  title: string
  subtitle?: string
} {
  if (!value) {
    return {title: blockTitle}
  }

  const filteredValue = filterValidBlocks(value)
  const textBlock = findTitleBlock(filteredValue)

  const title = textBlock ? getBlockText(textBlock) : blockTitle

  return {
    title,
    subtitle: title === blockTitle ? undefined : blockTitle,
  }
}

/**
 * Filters out invalid blocks, specifically "normal" styled blocks
 * that contain only empty text.
 *
 * @param blocks - An array of Portable Text blocks.
 * @returns A filtered array without empty "normal" blocks.
 */
export function filterValidBlocks(blocks: PortableTextBlock[]): PortableTextBlock[] {
  return Array.isArray(blocks)
    ? blocks.filter(
        (block) =>
          block._type !== 'block' ||
          block.style !== 'normal' ||
          (isPortableTextTextBlock(block) &&
            block.children.some((child) => String(child.text).trim() !== '')),
      )
    : []
}

/**
 * Finds the first valid block that can be used as a title.
 * Prioritizes blocks with styles 'h1', 'h2', 'h3'.
 * If none are found, it checks if the first block is a valid text block.
 * If not, it falls back to the first valid "normal" block.
 *
 * @param blocks - A filtered array of Portable Text blocks.
 * @returns The first valid block or `undefined` if none found.
 */
export function findTitleBlock(blocks: PortableTextBlock[]): PortableTextBlock | undefined {
  return (
    blocks.find(
      (block) => block._type === 'block' && ['h1', 'h2', 'h3'].includes(String(block.style)),
    ) ||
    (isPortableTextTextBlock(blocks[0])
      ? blocks[0]
      : blocks.find(
          (block) => block._type === 'block' && ['normal'].includes(String(block.style)),
        ) || undefined)
  )
}

/**
 * Extracts and joins the text content of a Portable Text block.
 *
 * @param block - A Portable Text block.
 * @returns The concatenated text content of the block.
 */
export function getBlockText(block: PortableTextBlock): string {
  return isPortableTextTextBlock(block) ? block.children.map((child) => child.text).join('') : ''
}
