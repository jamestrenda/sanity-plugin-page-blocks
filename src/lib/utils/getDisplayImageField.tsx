import {defineField, FieldDefinition} from 'sanity'

import {CustomImageType} from '../../types'
import {getImageField} from './getImageField'

/**
 * Generates an image field configuration for a display image in Sanity schema.
 *
 * This function is specifically designed for images where captions are not required
 * (e.g., hero images, background images). It utilizes `getImageField` but excludes
 * the `caption` field.
 *
 * @param {Exclude<CustomImageType, false>} options - Configuration options for the display image field.
 *   - Additional `customFields` can be specified to extend the image schema.
 * @returns {ImageField} - The configured display image field with the `caption` field excluded.
 */
export function getDisplayImage(
  options: Omit<Exclude<CustomImageType, false>, 'caption'> = {},
): FieldDefinition {
  // Use getImageField and filter out the 'caption' field specifically for display images
  const imageFieldConfig = getImageField(options)

  // Filter out the 'caption' field for display images
  const filteredFields = imageFieldConfig.fields.filter((field) => field.name !== 'caption')

  return defineField({...imageFieldConfig, fields: filteredFields})
}
