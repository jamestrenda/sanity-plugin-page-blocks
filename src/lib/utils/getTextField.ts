import {BlockDefinition, BlockStyleDefinition, defineField} from 'sanity'

import {TextType} from '../../types'
import {createFieldConfig} from './createSchema'
import {getPortableTextBlocks, portableTextBlocks} from './getPortableTextBlocks'

export function getTextField(
  options?: TextType,
  blockDef: Omit<BlockDefinition, 'type' | 'name'> = {
    styles: portableTextBlocks.styles ?? [],
    marks: {
      decorators: portableTextBlocks.marks?.decorators ?? [],
      annotations: portableTextBlocks.marks?.annotations ?? [],
    },
  },
  description?: string,
) {
  const text = options

  const defaultTextFieldBase = defineField({
    name: 'text',
    title: 'Text',
    description,
    type: 'array',
    of: [],
  })

  // Determine the text field dynamically
  if (text === false) return defaultTextFieldBase

  // use the default field if the user doesn't pass any customizations for text
  if (!text) {
    return defineField({
      ...defaultTextFieldBase,
      of: [...getPortableTextBlocks(blockDef)],
      validation: (Rule) => Rule.required(),
    })
  }

  // use a simple string field if the text is a string
  if (text.type === 'string') {
    return defineField({
      name: defaultTextFieldBase.name,
      title: defaultTextFieldBase.title,
      description,
      type: 'string',
      validation: text.validation ? text.validation : (Rule) => Rule.required(),
      ...createFieldConfig(text),
    })
  }

  // Merge user-defined properties with defaults
  return defineField({
    ...defaultTextFieldBase,
    of: [
      ...getPortableTextBlocks({
        styles: text.styles ?? blockDef.styles,
        lists: text.lists ?? [],
        marks: {
          decorators: text?.marks?.decorators ?? undefined,
          annotations: text?.marks?.annotations ?? undefined,
        },
      }),
      ...(text.of ?? []), // Add any additional block types defined by the user
    ],
    validation: text.validation ?? defaultTextFieldBase.validation,
    ...createFieldConfig(text),
  })
}
