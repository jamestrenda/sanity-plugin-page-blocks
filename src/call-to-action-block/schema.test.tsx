import {ObjectDefinition, SchemaTypeDefinition} from 'sanity'
import {describe, expect, it} from 'vitest'
import {callToActionBlock} from '.'
import {sanitizeSchema} from '../tests/lib/utils'

it('call-to-action-block generates default schema', () => {
  const plugin = callToActionBlock() // or pass config here
  const types = plugin?.schema?.types as SchemaTypeDefinition[]
  const schema = types[0] as ObjectDefinition
  expect(sanitizeSchema(schema)).toMatchSnapshot('Default callToAction schema')
})

it('call-to-action-block generates default schema without title field when title set to false', () => {
  const plugin = callToActionBlock({
    title: false,
  }) // or pass config here
  const types = plugin?.schema?.types as SchemaTypeDefinition[]
  const schema = types[0] as ObjectDefinition
  expect(sanitizeSchema(schema)).toMatchSnapshot('callToAction schema without title')
})

it('call-to-action-block generates default schema without text field when text set to false', () => {
  const plugin = callToActionBlock({
    text: false,
  }) // or pass config here
  const types = plugin?.schema?.types as SchemaTypeDefinition[]
  const schema = types[0] as ObjectDefinition
  expect(sanitizeSchema(schema)).toMatchSnapshot('callToAction schema without text')
})

describe('call-to-action-block metadata', () => {
  const plugin = callToActionBlock()
  const types = plugin?.schema?.types as SchemaTypeDefinition[]
  const schema = types[0] as ObjectDefinition

  it('have the correct name', () => {
    const schemaName = schema.name

    expect(schemaName).toEqual('callToActionBlock')
  })

  it('should have the correct title', () => {
    const schemaTitle = schema.title

    expect(schemaTitle).toEqual('Call to Action')
  })
})
