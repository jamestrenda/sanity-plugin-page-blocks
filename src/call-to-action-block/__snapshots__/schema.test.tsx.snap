// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`call-to-action-block generates default schema > Default callToAction schema 1`] = `
{
  "fields": [
    {
      "name": "title",
      "title": "Title",
      "type": "string",
    },
    {
      "description": "The main text of the call to action block.",
      "name": "text",
      "of": [
        {
          "lists": [
            {
              "title": "Bullet",
              "value": "bullet",
            },
            {
              "title": "Number",
              "value": "number",
            },
          ],
          "marks": {
            "decorators": [
              {
                "title": "Strong",
                "value": "strong",
              },
              {
                "title": "Emphasis",
                "value": "em",
              },
              {
                "title": "Underline",
                "value": "underline",
              },
              {
                "title": "Strike",
                "value": "strike-through",
              },
              {
                "title": "Code",
                "value": "code",
              },
            ],
          },
          "styles": [
            {
              "title": "Normal",
              "value": "normal",
            },
            {
              "title": "H2",
              "value": "h2",
            },
          ],
          "type": "block",
        },
      ],
      "title": "Text",
      "type": "array",
    },
    {
      "fields": [
        {
          "name": "file",
          "options": {
            "hotspot": true,
          },
          "title": "File",
          "type": "image",
        },
        {
          "description": "Override the alt text for this asset",
          "name": "altText",
          "title": "Alt Text",
          "type": "string",
        },
      ],
      "name": "customImage",
      "title": "Image",
      "type": "object",
    },
    {
      "description": "Add a call-to-action button or link to direct visitors to a specific page.",
      "name": "actions",
      "of": [
        {
          "fields": [
            {
              "fields": [
                {
                  "name": "text",
                  "title": "Text",
                  "type": "string",
                },
                {
                  "description": "Link to an internal reference, external or relative URL, or a file.",
                  "name": "to",
                  "of": [
                    {
                      "fields": [
                        {
                          "fields": [
                            {
                              "name": "url",
                              "title": "URL",
                              "type": "url",
                            },
                            {
                              "initialValue": true,
                              "name": "newWindow",
                              "title": "Open in a new window?",
                              "type": "boolean",
                            },
                          ],
                          "name": "link",
                          "title": "External Link",
                          "type": "object",
                        },
                      ],
                      "name": "external",
                      "title": "External Link",
                      "type": "object",
                    },
                    {
                      "fields": [
                        {
                          "description": "Example: /a-relative-url",
                          "name": "url",
                          "title": "Relative URL",
                          "type": "string",
                        },
                      ],
                      "name": "relative",
                      "title": "Relative URL",
                      "type": "object",
                    },
                    {
                      "fields": [
                        {
                          "fields": [
                            {
                              "description": "Select a file to link to",
                              "name": "file",
                              "title": "File",
                              "type": "file",
                            },
                          ],
                          "name": "link",
                          "title": "Media",
                          "type": "object",
                        },
                      ],
                      "name": "media",
                      "title": "Media Link",
                      "type": "object",
                    },
                  ],
                  "title": "To (Choose one)",
                  "type": "array",
                },
              ],
              "name": "action",
              "title": "Action",
              "type": "object",
            },
          ],
          "name": "action",
          "title": "Action",
          "type": "object",
        },
      ],
      "title": "Actions",
      "type": "array",
    },
  ],
  "fieldsets": [],
  "groups": [],
  "name": "callToActionBlock",
  "title": "Call to Action",
  "type": "object",
}
`;

exports[`call-to-action-block generates default schema without text field when text set to false > callToAction schema without text 1`] = `
{
  "fields": [
    {
      "name": "title",
      "title": "Title",
      "type": "string",
    },
    {
      "fields": [
        {
          "name": "file",
          "options": {
            "hotspot": true,
          },
          "title": "File",
          "type": "image",
        },
        {
          "description": "Override the alt text for this asset",
          "name": "altText",
          "title": "Alt Text",
          "type": "string",
        },
      ],
      "name": "customImage",
      "title": "Image",
      "type": "object",
    },
    {
      "description": "Add a call-to-action button or link to direct visitors to a specific page.",
      "name": "actions",
      "of": [
        {
          "fields": [
            {
              "fields": [
                {
                  "name": "text",
                  "title": "Text",
                  "type": "string",
                },
                {
                  "description": "Link to an internal reference, external or relative URL, or a file.",
                  "name": "to",
                  "of": [
                    {
                      "fields": [
                        {
                          "fields": [
                            {
                              "name": "url",
                              "title": "URL",
                              "type": "url",
                            },
                            {
                              "initialValue": true,
                              "name": "newWindow",
                              "title": "Open in a new window?",
                              "type": "boolean",
                            },
                          ],
                          "name": "link",
                          "title": "External Link",
                          "type": "object",
                        },
                      ],
                      "name": "external",
                      "title": "External Link",
                      "type": "object",
                    },
                    {
                      "fields": [
                        {
                          "description": "Example: /a-relative-url",
                          "name": "url",
                          "title": "Relative URL",
                          "type": "string",
                        },
                      ],
                      "name": "relative",
                      "title": "Relative URL",
                      "type": "object",
                    },
                    {
                      "fields": [
                        {
                          "fields": [
                            {
                              "description": "Select a file to link to",
                              "name": "file",
                              "title": "File",
                              "type": "file",
                            },
                          ],
                          "name": "link",
                          "title": "Media",
                          "type": "object",
                        },
                      ],
                      "name": "media",
                      "title": "Media Link",
                      "type": "object",
                    },
                  ],
                  "title": "To (Choose one)",
                  "type": "array",
                },
              ],
              "name": "action",
              "title": "Action",
              "type": "object",
            },
          ],
          "name": "action",
          "title": "Action",
          "type": "object",
        },
      ],
      "title": "Actions",
      "type": "array",
    },
  ],
  "fieldsets": [],
  "groups": [],
  "name": "callToActionBlock",
  "title": "Call to Action",
  "type": "object",
}
`;

exports[`call-to-action-block generates default schema without title field when title set to false > callToAction schema without title 1`] = `
{
  "fields": [
    {
      "description": "The main text of the call to action block.",
      "name": "text",
      "of": [
        {
          "lists": [
            {
              "title": "Bullet",
              "value": "bullet",
            },
            {
              "title": "Number",
              "value": "number",
            },
          ],
          "marks": {
            "decorators": [
              {
                "title": "Strong",
                "value": "strong",
              },
              {
                "title": "Emphasis",
                "value": "em",
              },
              {
                "title": "Underline",
                "value": "underline",
              },
              {
                "title": "Strike",
                "value": "strike-through",
              },
              {
                "title": "Code",
                "value": "code",
              },
            ],
          },
          "styles": [
            {
              "title": "Normal",
              "value": "normal",
            },
            {
              "title": "H2",
              "value": "h2",
            },
          ],
          "type": "block",
        },
      ],
      "title": "Text",
      "type": "array",
    },
    {
      "fields": [
        {
          "name": "file",
          "options": {
            "hotspot": true,
          },
          "title": "File",
          "type": "image",
        },
        {
          "description": "Override the alt text for this asset",
          "name": "altText",
          "title": "Alt Text",
          "type": "string",
        },
      ],
      "name": "customImage",
      "title": "Image",
      "type": "object",
    },
    {
      "description": "Add a call-to-action button or link to direct visitors to a specific page.",
      "name": "actions",
      "of": [
        {
          "fields": [
            {
              "fields": [
                {
                  "name": "text",
                  "title": "Text",
                  "type": "string",
                },
                {
                  "description": "Link to an internal reference, external or relative URL, or a file.",
                  "name": "to",
                  "of": [
                    {
                      "fields": [
                        {
                          "fields": [
                            {
                              "name": "url",
                              "title": "URL",
                              "type": "url",
                            },
                            {
                              "initialValue": true,
                              "name": "newWindow",
                              "title": "Open in a new window?",
                              "type": "boolean",
                            },
                          ],
                          "name": "link",
                          "title": "External Link",
                          "type": "object",
                        },
                      ],
                      "name": "external",
                      "title": "External Link",
                      "type": "object",
                    },
                    {
                      "fields": [
                        {
                          "description": "Example: /a-relative-url",
                          "name": "url",
                          "title": "Relative URL",
                          "type": "string",
                        },
                      ],
                      "name": "relative",
                      "title": "Relative URL",
                      "type": "object",
                    },
                    {
                      "fields": [
                        {
                          "fields": [
                            {
                              "description": "Select a file to link to",
                              "name": "file",
                              "title": "File",
                              "type": "file",
                            },
                          ],
                          "name": "link",
                          "title": "Media",
                          "type": "object",
                        },
                      ],
                      "name": "media",
                      "title": "Media Link",
                      "type": "object",
                    },
                  ],
                  "title": "To (Choose one)",
                  "type": "array",
                },
              ],
              "name": "action",
              "title": "Action",
              "type": "object",
            },
          ],
          "name": "action",
          "title": "Action",
          "type": "object",
        },
      ],
      "title": "Actions",
      "type": "array",
    },
  ],
  "fieldsets": [],
  "groups": [],
  "name": "callToActionBlock",
  "title": "Call to Action",
  "type": "object",
}
`;
